Add Import/Export mechanism + code migration from 4ever codebase

Gbury opened the pull request:
Note: this is still WIP.

This adds the possibility to import and export events from files in toml format. Additionally, it adds a bit of logging to make it easier to debug things.

But2ene commented on src/lib/results.ml:
```diff
@@ -0,0 +1,165 @@
+  let conv = Conv.mk [nullable int] CCFun.id in
+  CCOption.get_or ~default:0 @@
+  State.query_one_where ~st ~conv ~p:[int; int; int]
+    {| SELECT SUM(results.points)
+       FROM results JOIN competitions ON results.competition=competitions.id
+       WHERE results.dancer = ? AND results.role = ? AND competitions.category = ? |}
```
a beautiful join. Do you want a position as a data scientist?

Gbury replied:
I think I'll stay on the compilation part of the line for now, :p (I'm still curious/worried about how exactly that query is actually compiled, but for now it'll do, ^^)

Gbury commented on src/backend/types.ml:
```diff
@@ -281,6 +281,8 @@ module Competition = struct
     name : string;
     kind : Kind.t;
     category : Category.t;
+    leaders : int;
+    follows : int;
```
changed to `leaders_count` and `followers_count`

Gbury commented on src/lib/competition.ml:
```diff
@@ -60,14 +74,31 @@ let from_event st event_id =
-    {| INSERT INTO competitions (event, name, kind, category) VALUES (?,?,?,?) |}
-    event_id name (Kind.to_int kind) (Category.to_int category);
+  State.insert ~st ~ty:[ int; text; int; int; int; int ]
+    {| INSERT INTO competitions (event, name, kind, category, num_leaders, num_follows)
+       VALUES (?,?,?,?,?,?) |}
```
fixed

Gbury commented on src/lib/competition.mli:
```diff
@@ -29,6 +29,20 @@ val kind : t -> Kind.t
+    Note that in some cases (mostly for old competitions), this information
+    may be missing and therefore this will return [0]. *)
+
+val check_divs : t -> bool
+(** Should the competition check the divisoin of participants ? This is only
+    set to [false] for old competitions during the introduction of the SCR4T
+    competitive point system. *)
```
I don't agree: by default we should check divs to ensure that no errors has been done (for instance, a new dancer mistakenly created for an intermediate dancer could be caught by this check, or also if a competition organizer did not check the divisions of competitors), basically, the check is needed to ensure that the rues of the SCR4T have been followed, and it's better for it to be automated (and possibly redundant).

That being said, that means that we will need a system to store and take into account petitions (i.e. individuals exceptions to the check). Until that is implemented, the process will be to have a manual verification of the competitions where the check does not pass, and then to disable checking for this competition (and later add the relevant petition so that the check can be reinstated).

Gbury commented on src/lib/import.ml:
```diff
@@ -0,0 +1,251 @@
+(* Event *)
+(* ************************************************************************* *)
+
+let import_event ~st ~subst ~t =
+  let open Otoml in
+  let+ name = find_result t get_string ["name"] in
+  let+ start_date = find_result t Date.of_toml ["start_date"] in
+  let+ end_date = find_result t Date.of_toml ["end_date"] in
+  let event_id = Event.create st name ~start_date ~end_date in
+  let+ t = find_result t Otoml.get_value ["comps"] in
+  import_comps ~st ~subst ~t ~event_id
```
Indeed, if you look at how the import function is called in `backend/main.ml` you'll see that it is wrapped in a call to `State.atomically` which implements a transaction, which means that if the import fails, then the changes will not be comitted.
We can discuss whether the `lib/import.ml` module should enforce that call to `State.atomically` though (rather than letting its caller handle that).

Gbury commented:
I fixed the comments and bugs, and added some of the old competitions results, so that it's easy to create a test database (see `HACKING.md`), in the future we may automate the test db creation, ^^

@But2ene you can take another look if you want (since I've added a few code since you reviewed), but I'll try and merge this soon (this weekend probably) so that the other PRs can be rebased and merged, :)
